packages/adapters-library/src/adapters/stargate/products/farm-v2/stargateFarmV2Adapter.ts

```diff
@@ -0,0 +1,374 @@
+      this.provider,
+    )
+
+    const [rewardTokenAddresses, rewardAmounts] =
+      await rewarderContract.getRewards(protocolTokenAddress, userAddress, {
+        blockTag: blockNumber,
+      })
+
+    return filterMapSync(rewardTokenAddresses, (rewardTokenAddress, i) => {
+      const rewardAmount = rewardAmounts[i]!
+
+      if (!rewardAmount) {
+        return undefined
+      }
+
+      const rewardToken = rewardTokens.find(
+        (rt) => rt.address === rewardTokenAddress,
+      )
+
+      if (!rewardToken) {
+        throw Error('Missing reward token from Metadata')
+      }
+
+      return {
+        ...rewardToken,
+        type: TokenType.UnderlyingClaimable,
+        balanceRaw: rewardAmount,
+      }
+    })
+  }
+
+  async getWithdrawals(input: GetEventsInput): Promise<MovementsByBlock[]> {
+    return await this.getMovements({
+      ...input,
+      filterType: 'withdrawal',
+    })
+  }
+
+  async getDeposits(input: GetEventsInput): Promise<MovementsByBlock[]> {
+    return await this.getMovements({
+      ...input,
+      filterType: 'deposit',
+    })
+  }
+
+  private async getMovements({
+    protocolTokenAddress,
+    fromBlock,
+    toBlock,
+    userAddress,
+    filterType,
+  }: GetEventsInput & {
+    filterType: 'deposit' | 'withdrawal'
+  }): Promise<MovementsByBlock[]> {
+    const { stargateStakingAddress } = staticChainDataV2[this.chainId]!
+
+    const protocolToken =
+      await this.getProtocolTokenByAddress(protocolTokenAddress)
+
+    const lpStakingContract = StargateStaking__factory.connect(
+      stargateStakingAddress,
+      this.provider,
+    )
+
+    const filter =
+      filterType === 'deposit'
+        ? lpStakingContract.filters.Deposit(
+            protocolTokenAddress,
+            undefined,
+            userAddress, // Address that receives the tokens (to)
+            undefined,
+          )
+        : lpStakingContract.filters.Withdraw(
+            protocolTokenAddress,
+            userAddress, // Address from which the tokens are withdrawn (from)
+            undefined,
+            undefined,
+          )
+
+    const events = await lpStakingContract.queryFilter(
+      filter,
+      fromBlock,
+      toBlock,
+    )
+
+    return events.map((event) => {
+      const { amount } = event.args!
+
+      return {
+        protocolToken: {
+          address: protocolToken.address,
+          name: protocolToken.name,
+          symbol: protocolToken.symbol,
+          decimals: protocolToken.decimals,
+        },
+        blockNumber: event.blockNumber,
+        transactionHash: event.transactionHash,
+        tokens: [
+          {
+            address: protocolToken.address,
+            name: protocolToken.name,
+            symbol: protocolToken.symbol,
+            decimals: protocolToken.decimals,
+            type: TokenType.Underlying,
+            blockNumber: event.blockNumber,
+            balanceRaw: amount,
+          },
+        ],
+      }
+    })
+  }
```

Comment:
This is not a blocker but i  think you might have been able to use the transfer movements of the protocol token

```
    const movements = await this.helpers.getErc20Movements({
      protocolToken: underlyingLpToken,

      filter: {
        fromBlock,
        toBlock,
        from: userAddress,
        to: protocolTokenAddress,
      },
    })
```

Reply:
We have a helper

Reply:
but please dont bother to update its not worth it

Reply:
100% worth it if we have a helper. I'll update.